documentation.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2012-2022 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENSE file for details.
+
+package cmd
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+
+	"github.com/juju/gnuflag"
+)
+
+var doc string = `
+This command generates a markdown formatted document with all the commands, their descriptions, arguments, and examples.
+`
+
+type documentationCommand struct {
+	CommandBase
+	super   *SuperCommand
+	out     string
+	noIndex bool
+}
+
+func newDocumentationCommand(s *SuperCommand) *documentationCommand {
+	return &documentationCommand{super: s}
+}
+
+func (c *documentationCommand) Info() *Info {
+	return &Info{
+		Name:    "documentation",
+		Args:    "--out <target-file> --noindex",
+		Purpose: "Generate the documentation for all commands",
+		Doc:     doc,
+	}
+}
+
+// SetFlags adds command specific flags to the flag set.
+func (c *documentationCommand) SetFlags(f *gnuflag.FlagSet) {
+	f.StringVar(&c.out, "out", "", "Documentation output file")
+	f.BoolVar(&c.noIndex, "noindex", false, "Do not generate the commands index")
+}
+
+func (c *documentationCommand) Run(ctx *Context) error {
+	var writer *bufio.Writer
+	if c.out != "" {
+		f, err := os.Create(c.out)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		writer = bufio.NewWriter(f)
+	} else {
+		writer = bufio.NewWriter(ctx.Stdout)
+	}
+	return c.dumpEntries(writer)
+}
+
+func (c *documentationCommand) dumpEntries(writer *bufio.Writer) error {
+	if len(c.super.subcmds) == 0 {
+		return nil
+	}
+
+	// sort the commands
+	sorted := make([]string, len(c.super.subcmds))
+	i := 0
+	for k := range c.super.subcmds {
+		sorted[i] = k
+		i++
+	}
```

Comment:
```suggestion
	sorted := make([]string, 0, len(c.super.subcmds))
	for k := range c.super.subcmds {
		sorted = append(sorted, k)
	}
```

Reply:
If the length of a slice is known beforehand, it is better to preallocate with the length indicated in the `make`, that avoid additional memory overhead. In this case, the size is not large enough to be a problem. I will leave it as it is right now.

Reply:
Well that's also what capacity does, you're saying here, have the initial capacity of the sub-commands length, only when you grow past that, allocate more (I think the strategy is to double). Either way.

Reply:
I think `append` will use copy functions and if you know the size you should probably avoid that (https://go.dev/blog/slices) the performance improvement is not relevant in this case anyway.